### PR TITLE
Move the NCCL module to `cupy_backends.cuda.libs`

### PR DESCRIPTION
**Pull request by leofang**
Prepare for #4873.

The change is a three-step one:
1. Move and break the header `cupy/cuda/cupy_nccl.h` into 3 pieces (for CUDA, HIP, and RTD)
2. Move `cupy/cuda/nccl.{pxd, pyx}` to `cupy_backends/cuda/libs`
3. Add `cupy/cuda/nccl.py` to import the moved module

### CODE REVIEW

File: cupy/cuda/nccl.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+"""
+NCCL Wrapper
+
+Use `cupy_backends.cuda.libs.nccl` directly in CuPy codebase.
```
Comment (kmaehashi) [inline cupy/cuda/nccl.py:4]: `cupyx/_runtime.py` uses `cupy.cuda.nccl`, so it should be updated.

Comment (leofang) [reply cupy/cuda/nccl.py:4]: Good catch, thanks! Fixed in 437ee18.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by leofang** [comment]
Note: we still need to do some cleanup for the NCCL codes (ex: #3677), but I prefer to not touch it in this PR. It's not urgent I think.

**Comment by leofang** [comment]
Hmm...I thought I fixed it. Let me check.

**Comment by leofang** [comment]
The fix 67289a7 is due to https://github.com/cupy/cupy/pull/4920#discussion_r597843135.

**Comment by leofang** [comment]
Jenkins, test this please

File: cupy/cuda/__init__.py [hunk 2]
```diff
@@ -51,10 +51,7 @@ def __init__(self, name):
 except ImportError:
     thrust = _UnavailableModule('cupy.cuda.thrust')
 
-try:
-    from cupy.cuda import nccl  # NOQA
-except ImportError:
-    nccl = _UnavailableModule('cupy.cuda.nccl')
+from cupy.cuda import nccl  # NOQA
```
Comment (kmaehashi) [inline cupy/cuda/__init__.py:54]: This line needs to be removed.

Comment (leofang) [reply cupy/cuda/__init__.py]: Fixed in 9e8192b.

**Comment by leofang** [comment]
Jenkins, test this please

File: tests/cupy_tests/cuda_tests/test_nccl.py [hunk 3]
```diff
@@ -19,60 +20,60 @@ class TestNCCL(unittest.TestCase):
 
     @testing.gpu
     def test_single_proc_ring(self):
-        id = cuda.nccl.get_unique_id()
-        comm = cuda.nccl.NcclCommunicator(1, id, 0)
+        id = nccl.get_unique_id()
```
Comment (kmaehashi) [inline tests/cupy_tests/cuda_tests/test_nccl.py:23]: FYI (no need to revert this), you didn't have to change `cuda.nccl` to `nccl` just by adding `import cupy.cuda.nccl` 😄

**Review by kmaehashi — APPROVED** [review]

**Comment by leofang** [comment]
Thanks!

**Pull request merged**